collaborative-plonk/src/round4.rs

```diff
@@ -0,0 +1,240 @@
+        + MSMProvider<P::G2>,
+    P::ScalarField: mpc_core::traits::FFTPostProcessing,
+{
+    pub(super) driver: T,
+    pub(super) domains: Domains<P>,
+    pub(super) challenges: Round3Challenges<T, P>,
+    pub(super) proof: Round3Proof<P>,
+    pub(super) polys: FinalPolys<T, P>,
+    pub(super) data: PlonkData<T, P>,
+}
+pub(super) struct Round4Challenges<P: Pairing> {
+    pub(super) beta: P::ScalarField,
+    pub(super) gamma: P::ScalarField,
+    pub(super) alpha: P::ScalarField,
+    pub(super) xi: P::ScalarField,
+}
+impl<P: Pairing> Round4Challenges<P> {
+    fn new<T>(round3_challenges: Round3Challenges<T, P>, xi: P::ScalarField) -> Self
+    where
+        for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>,
+    {
+        Self {
+            beta: round3_challenges.beta,
+            gamma: round3_challenges.gamma,
+            alpha: round3_challenges.alpha,
+            xi,
+        }
+    }
+}
+
+pub(super) struct Round4Proof<P: Pairing> {
+    pub(super) commit_a: P::G1,
+    pub(super) commit_b: P::G1,
+    pub(super) commit_c: P::G1,
+    pub(super) commit_z: P::G1,
+    pub(super) commit_t1: P::G1,
+    pub(super) commit_t2: P::G1,
+    pub(super) commit_t3: P::G1,
+    pub(super) eval_a: P::ScalarField,
+    pub(super) eval_b: P::ScalarField,
+    pub(super) eval_c: P::ScalarField,
+    pub(super) eval_zw: P::ScalarField,
+    pub(super) eval_s1: P::ScalarField,
+    pub(super) eval_s2: P::ScalarField,
+}
+
+impl<P: Pairing> Round4Proof<P> {
+    fn new(
+        round3_proof: Round3Proof<P>,
+        eval_a: P::ScalarField,
+        eval_b: P::ScalarField,
+        eval_c: P::ScalarField,
+        eval_zw: P::ScalarField,
+        eval_s1: P::ScalarField,
+        eval_s2: P::ScalarField,
+    ) -> Self {
+        Self {
+            commit_a: round3_proof.commit_a,
+            commit_b: round3_proof.commit_b,
+            commit_c: round3_proof.commit_c,
+            commit_z: round3_proof.commit_z,
+            commit_t1: round3_proof.commit_t1,
+            commit_t2: round3_proof.commit_t2,
+            commit_t3: round3_proof.commit_t3,
+            eval_a,
+            eval_b,
+            eval_c,
+            eval_zw,
+            eval_s1,
+            eval_s2,
+        }
+    }
+}
+
+impl<T, P: Pairing> Round4<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+{
+    fn evaluate_poly(
+        driver: &mut T,
+        poly: &FieldShareVec<T, P>,
+        x: &P::ScalarField,
+    ) -> FieldShare<T, P> {
+        let mut res = FieldShare::<T, P>::default();
+        let mut x_pow = P::ScalarField::one();
+        for coeff in poly.clone().into_iter() {
+            let tmp = driver.mul_with_public(&x_pow, &coeff);
+            res = driver.add(&res, &tmp);
+            x_pow *= x;
+        }
+        res
+    }
```

Comment:
This evaluate poly method is pretty slow, can we not call down into arkworks for that somehow?

Reply:
done, added the evaluate_poly_public function to the FFT trait